### PR TITLE
Hard: incentive reward PR #2

### PR DESCRIPTION
**Pull request by denalimarsh**
- [x] Update /x/incentive keeper methods and types for hard supply/borrow/delegation rewards
- [x] Implement the hooks methods in `x/incentive` for supply/borrow (not delegation)
- [x] Implement reward accumulators for supply/borrow
- [x] Write tests for supply/borrow reward accumulators and reward syncing for supply/borrow rewards

Base branch is set to `dm-hard-rewards-incentive` for easier review

### CODE REVIEW

File: x/incentive/types/keys.go [hunk 1]
```diff
@@ -17,11 +17,20 @@ const (
...
 )
 
+// TODO: Could store previous reward time accruals under the same key indexed by type ('usd_minting', 'hard_supply', etc.)
+
 // Key Prefixes
 var (
-	ClaimKeyPrefix  = []byte{0x01} // prefix for keys that store claims
-	BlockTimeKey    = []byte{0x02} // prefix for key that stores the blocktime
-	RewardFactorKey = []byte{0x03} // prefix for key that stores reward factors
+	USDXMintingClaimKeyPrefix                       = []byte{0x01} // prefix for keys that store USDX minting claims
```
Comment (karzak) [inline x/incentive/types/keys.go:24]: Store structure looks fine. Generically, for each incentive, there's:
* A key that stores the claims
* A key that stores the global accumulator value
* A key that stores the last time the accumulator was updated.

One option would be to combine the last two keys into a struct and just update the struct whenever the accumulator is updated, but I'll leave that as an implementation detail and don't think it's important to change.

Comment (denalimarsh) [reply x/incentive/types/keys.go:26]: Added a TODO to refactor towards this structure

**Comment by denalimarsh** [comment]
On this branch the Hard module doesn't accumulate supply-side interest and deposits don't increment supplied coins, so I had to add some scaffolding back in to make things work. In order to do proper e2e testing, we'll need to rebase this onto master so the Hard module includes the latest changes. Since the surface area is already so large I think it makes sense to continue with the implementation then add another story for the master integration and expanding the test suite once this set of stories is completed.

**Review by karzak — CHANGES_REQUESTED** [review]

File: x/incentive/keeper/rewards.go [hunk 2]
```diff
@@ -7,60 +7,140 @@ import (
...
+	}
+	timeElapsed := CalculateTimeElapsed(rewardPeriod, ctx.BlockTime(), previousAccrualTime)
+	if timeElapsed.IsZero() {
+		return nil
+	}
+	if rewardPeriod.RewardsPerSecond.Amount.IsZero() {
+		k.SetPreviousHardBorrowRewardAccrualTime(ctx, rewardPeriod.CollateralType, ctx.BlockTime())
+		return nil
+	}
+	totalBorrowedCoins, _ := k.hardKeeper.GetBorrowedCoins(ctx)
```
[the remark this answers is not in the document]
Comment (denalimarsh) [reply x/incentive/keeper/rewards.go]: Updated in https://github.com/Kava-Labs/kava/pull/757/commits/613f5ab4e8ac9dadf019a2a18cd160b96eb05227

File: x/incentive/keeper/rewards.go [hunk 3]
```diff
@@ -7,60 +7,140 @@ import (
...
+	timeElapsed := CalculateTimeElapsed(rewardPeriod, ctx.BlockTime(), previousAccrualTime)
+	if timeElapsed.IsZero() {
+		return nil
+	}
+	if rewardPeriod.RewardsPerSecond.Amount.IsZero() {
+		k.SetPreviousHardSupplyRewardAccrualTime(ctx, rewardPeriod.CollateralType, ctx.BlockTime())
+		return nil
+	}
+
+	totalSuppliedCoins, _ := k.hardKeeper.GetSuppliedCoins(ctx)
```
[the remark this answers is not in the document]
Comment (denalimarsh) [reply x/incentive/keeper/rewards.go]: Fixed in https://github.com/Kava-Labs/kava/pull/757/commits/613f5ab4e8ac9dadf019a2a18cd160b96eb05227

File: x/incentive/keeper/rewards.go [hunk 4]
```diff
@@ -70,34 +150,34 @@ func (k Keeper) InitializeClaim(ctx sdk.Context, cdp cdptypes.CDP) {
...
-	k.SetClaim(ctx, claim)
+	k.SetUSDXMintingClaim(ctx, claim)
 }
 
-// SynchronizeReward updates the claim object by adding any accumulated rewards and updating the reward index value.
+// SynchronizeUSDXMintingReward updates the claim object by adding any accumulated rewards and updating the reward index value.
 // this should be called before a cdp is modified, immediately after the 'SynchronizeInterest' method is called in the cdp module
-func (k Keeper) SynchronizeReward(ctx sdk.Context, cdp cdptypes.CDP) {
+func (k Keeper) SynchronizeUSDXMintingReward(ctx sdk.Context, cdp cdptypes.CDP) {
 	_, found := k.GetRewardPeriod(ctx, cdp.Type)
```
[the remark this answers is not in the document]
Comment (denalimarsh) [reply x/incentive/keeper/rewards.go]: Fixed in https://github.com/Kava-Labs/kava/pull/757/commits/41bafb70f0902c8fb50a7338e928beae901ae82e

File: x/hard/types/expected_keepers.go [hunk 5]
```diff
@@ -48,8 +48,8 @@ type AuctionKeeper interface {
 
 // HARDHooks event hooks for other keepers to run code in response to HARD modifications
 type HARDHooks interface {
```
[the remark this answers is not in the document]
Comment (denalimarsh) [reply x/hard/types/expected_keepers.go:50]: Updated in latest

File: x/incentive/keeper/rewards.go [hunk 6]
```diff
@@ -7,60 +7,140 @@ import (
...
-	k.SetRewardFactor(ctx, rewardPeriod.CollateralType, newRewardFactor)
-	k.SetPreviousAccrualTime(ctx, rewardPeriod.CollateralType, ctx.BlockTime())
+	k.SetUSDXMintingRewardFactor(ctx, rewardPeriod.CollateralType, newRewardFactor)
+	k.SetPreviousUSDXMintingAccrualTime(ctx, rewardPeriod.CollateralType, ctx.BlockTime())
 	return nil
 }
 
-// InitializeClaim creates or updates a claim such that no new rewards are accrued, but any existing rewards are not lost.
+// AccumulateHardBorrowRewards updates the rewards accumulated for the input reward period
+func (k Keeper) AccumulateHardBorrowRewards(ctx sdk.Context, rewardPeriod types.RewardPeriod) error {
```
[the remark this answers is not in the document]
Comment (denalimarsh) [reply x/incentive/keeper/rewards.go:48]: added in https://github.com/Kava-Labs/kava/pull/757/commits/8f528ea4454d797790b875c9c9926a6ba97a3f33

File: x/incentive/keeper/rewards.go [hunk 7]
```diff
@@ -108,19 +184,202 @@ func (k Keeper) SynchronizeReward(ctx sdk.Context, cdp cdptypes.CDP) {
...
+
+	for _, coin := range deposit.Amount {
+		supplyFactor, found := k.GetHardSupplyRewardFactor(ctx, coin.Denom)
+		if !found {
+			continue
+		}
+
+		supplyIndex, hasSupplyRewardIndex := claim.HasSupplyRewardIndex(coin.Denom)
+		if !hasSupplyRewardIndex {
+			continue
```
Comment (karzak) [inline x/incentive/keeper/rewards.go:249]: Just as a note, I don't think this should ever happen. Might be worth adding a print statement or something here on testnet and seeing if it's ever hit.

File: x/incentive/keeper/rewards.go [hunk 8]
```diff
@@ -108,19 +184,202 @@ func (k Keeper) SynchronizeReward(ctx sdk.Context, cdp cdptypes.CDP) {
...
+
+		userRewardFactor := claim.SupplyRewardIndexes[supplyIndex].RewardFactor
+		rewardsAccumulatedFactor := supplyFactor.Sub(userRewardFactor)
+		if rewardsAccumulatedFactor.IsZero() {
+			continue
+		}
+		claim.SupplyRewardIndexes[supplyIndex].RewardFactor = supplyFactor
+
+		newRewardsAmount := rewardsAccumulatedFactor.Mul(deposit.Amount.AmountOf(coin.Denom).ToDec()).RoundInt()
+		if newRewardsAmount.IsZero() {
```
Comment (karzak) [inline x/incentive/keeper/rewards.go:260]:
```suggestion
if newRewardsAmount.IsZero() || newRewardsAmount.IsNegative() {
```

sanity check

File: x/incentive/keeper/rewards.go [hunk 9]
```diff
@@ -108,19 +184,202 @@ func (k Keeper) SynchronizeReward(ctx sdk.Context, cdp cdptypes.CDP) {
...
+
+	k.SetHardLiquidityProviderClaim(ctx, claim)
+}
+
+// InitializeHardBorrowReward initializes the borrow-side of a hard liquidity provider claim
+// by creating the claim and setting the borrow reward factor index
+func (k Keeper) InitializeHardBorrowReward(ctx sdk.Context, borrow hardtypes.Borrow) {
+	claim, claimFound := k.GetHardLiquidityProviderClaim(ctx, borrow.Borrower)
+	if !claimFound {
+		return
```
Comment (karzak) [inline x/incentive/keeper/rewards.go:276]: Imagine this scenario:

1. User deposits coinA, which we don't have incentives for - a claim will not be created.
2. User borrows coinB, which we do have incentives for - claim doesn't exist!

Again, if I have the state-machine correct, it's not safe to return here. If a claim doesn't exist:

```
if !claimFound {
  claim = types.NewHardLiquidityProviderClaim(deposit.Depositor,
			sdk.NewCoin(types.HardLiquidityRewardDenom, sdk.ZeroInt()),
			nil, nil, nil)
}
```

File: x/incentive/keeper/rewards.go [hunk 10]
```diff
@@ -108,19 +184,202 @@ func (k Keeper) SynchronizeReward(ctx sdk.Context, cdp cdptypes.CDP) {
...
+	}
+
+	k.SetHardLiquidityProviderClaim(ctx, claim)
+}
+
+// UpdateHardSupplyIndexDenoms adds any new deposit denoms to the claim's supply reward index
+func (k Keeper) UpdateHardSupplyIndexDenoms(ctx sdk.Context, deposit hardtypes.Deposit) {
+	claim, found := k.GetHardLiquidityProviderClaim(ctx, deposit.Depositor)
+	if !found {
+		return
```
Comment (karzak) [inline x/incentive/keeper/rewards.go:341]: Same scenario as above (a user first supplies a coin for which there are no incentives) - we should still initialize a claim if one is not found. 

```
if !found {
  claim = types.NewHardLiquidityProviderClaim(deposit.Depositor,
			sdk.NewCoin(types.HardLiquidityRewardDenom, sdk.ZeroInt()),
			nil, nil, nil)
}
```

File: x/incentive/keeper/rewards.go [hunk 11]
```diff
@@ -108,19 +184,202 @@ func (k Keeper) SynchronizeReward(ctx sdk.Context, cdp cdptypes.CDP) {
...
+		_, hasIndex := claim.HasSupplyRewardIndex(coin.Denom)
+		if !hasIndex {
+			supplyFactor, foundSupplyFactor := k.GetHardSupplyRewardFactor(ctx, coin.Denom)
+			if !foundSupplyFactor {
+				supplyRewardIndexes = append(supplyRewardIndexes, types.NewRewardIndex(coin.Denom, supplyFactor))
+			}
+		}
+	}
+	claim.SupplyRewardIndexes = supplyRewardIndexes
+	k.SetHardLiquidityProviderClaim(ctx, claim)
```
Comment (karzak) [inline x/incentive/keeper/rewards.go:355]: In the extreme following edge case:

1. User supplies coinA, no incentives - claim doesn't get created.
2. User supplies coinB, no incentives - claim gets created above, but rewardIndexes will still be nil.

To avoid setting a claim in the store which shouldn't exist

```
if claim.SupplyRewardIndexes == []RewardIndexes{} { // not sure exact syntax, could use reflect.DeepEqual I believe
  return
} 
k.SetHardLiquidityProviderClaim(ctx, claim)
```

File: x/incentive/keeper/rewards.go [hunk 12]
```diff
@@ -108,19 +184,202 @@ func (k Keeper) SynchronizeReward(ctx sdk.Context, cdp cdptypes.CDP) {
...
+}
+
+// UpdateHardBorrowIndexDenoms adds any new borrow denoms to the claim's supply reward index
+func (k Keeper) UpdateHardBorrowIndexDenoms(ctx sdk.Context, borrow hardtypes.Borrow) {
+	claim, found := k.GetHardLiquidityProviderClaim(ctx, borrow.Borrower)
+	if !found {
+		return
```
Comment (karzak) [inline x/incentive/keeper/rewards.go:362]: Same as with `UpdateHardSupplyIndexDenoms`

```
if !found {
  claim = types.NewHardLiquidityProviderClaim(deposit.Depositor,
            sdk.NewCoin(types.HardLiquidityRewardDenom, sdk.ZeroInt()),
            nil, nil, nil)
}
```

File: x/incentive/keeper/rewards.go [hunk 13]
```diff
@@ -108,19 +184,202 @@ func (k Keeper) SynchronizeReward(ctx sdk.Context, cdp cdptypes.CDP) {
...
+		_, hasIndex := claim.HasBorrowRewardIndex(coin.Denom)
+		if !hasIndex {
+			borrowFactor, foundBorrowFactor := k.GetHardBorrowRewardFactor(ctx, coin.Denom)
+			if foundBorrowFactor {
+				borrowRewardIndexes = append(borrowRewardIndexes, types.NewRewardIndex(coin.Denom, borrowFactor))
+			}
+		}
+	}
+	claim.BorrowRewardIndexes = borrowRewardIndexes
+	k.SetHardLiquidityProviderClaim(ctx, claim)
```
Comment (karzak) [inline x/incentive/keeper/rewards.go:376]: Same edge case as above:


```
if claim.BorrowRewardIndexes == []RewardIndexes{} { // not sure exact syntax, could use reflect.DeepEqual I believe
  return
} 
k.SetHardLiquidityProviderClaim(ctx, claim)
```

**Review by karzak — CHANGES_REQUESTED** [review]
Only had time to review the hooks implementation, found a few edge cases. For each of these, please respond if you thnk I might have missed something - the state machine is fairly subtle.

Will continue reviewing the rest of the PR in the AM.

File: x/incentive/keeper/rewards.go [hunk 14]
```diff
@@ -108,19 +184,202 @@ func (k Keeper) SynchronizeReward(ctx sdk.Context, cdp cdptypes.CDP) {
...
+
+		userRewardFactor := claim.BorrowRewardIndexes[borrowIndex].RewardFactor
+		rewardsAccumulatedFactor := borrowFactor.Sub(userRewardFactor)
+		if rewardsAccumulatedFactor.IsZero() {
+			continue
+		}
+		claim.BorrowRewardIndexes[borrowIndex].RewardFactor = borrowFactor
+
+		newRewardsAmount := rewardsAccumulatedFactor.Mul(borrow.Amount.AmountOf(coin.Denom).ToDec()).RoundInt()
+		if newRewardsAmount.IsZero() {
```
Comment (karzak) [inline x/incentive/keeper/rewards.go:326]:
```suggestion
if newRewardsAmount.IsZero() || newRewardsAmount.IsNegative() {
```

sanity check

File: x/incentive/keeper/rewards.go [hunk 15]
```diff
@@ -108,19 +184,202 @@ func (k Keeper) SynchronizeReward(ctx sdk.Context, cdp cdptypes.CDP) {
...
+	k.SetUSDXMintingClaim(ctx, claim)
 	return
 }
 
+// InitializeHardSupplyReward initializes the supply-side of a hard liquidity provider claim
+// by creating the claim and setting the supply reward factor index
+func (k Keeper) InitializeHardSupplyReward(ctx sdk.Context, deposit hardtypes.Deposit) {
+	_, claimFound := k.GetHardLiquidityProviderClaim(ctx, deposit.Depositor)
+	if claimFound {
+		return
```
[the remark this answers is not in the document]
Comment (denalimarsh) [reply x/incentive/keeper/rewards.go]: Yeah, I agree. This solution will also maintain any unclaimed rewards the user might have. Fixed in https://github.com/Kava-Labs/kava/pull/757/commits/9ac39520b814e0cef5e304679a7ddbbcf94181d0

File: x/incentive/keeper/rewards.go (hunk 7 above)
Comment (denalimarsh) [reply x/incentive/keeper/rewards.go:247]: Added print log (with TODO so we remove it later) in https://github.com/Kava-Labs/kava/pull/757/commits/172ecc89fbea476da913136f65873d17a21de110

File: x/incentive/keeper/rewards.go [hunk 16]
```diff
@@ -108,19 +184,202 @@ func (k Keeper) SynchronizeReward(ctx sdk.Context, cdp cdptypes.CDP) {
...
+
+		supplyFactor, foundSupplyFactor := k.GetHardSupplyRewardFactor(ctx, coin.Denom)
+		if !foundSupplyFactor {
+			supplyFactor = sdk.ZeroDec()
+		}
+
+		supplyRewardIndexes = append(supplyRewardIndexes, types.NewRewardIndex(coin.Denom, supplyFactor))
+	}
+
+	claim, found := k.GetHardLiquidityProviderClaim(ctx, deposit.Depositor)
```
[the remark this answers is not in the document]
Comment (denalimarsh) [reply x/incentive/keeper/rewards.go:215]: Updated in https://github.com/Kava-Labs/kava/pull/757/commits/9ac39520b814e0cef5e304679a7ddbbcf94181d0

File: x/incentive/keeper/rewards.go (hunk 8 above)
Comment (denalimarsh) [reply x/incentive/keeper/rewards.go]: https://github.com/Kava-Labs/kava/pull/757/commits/5b82f9a1b6374e4973ea4c8769fe1937c0fcb9e6

File: x/incentive/keeper/rewards.go (hunk 14 above)
Comment (denalimarsh) [reply x/incentive/keeper/rewards.go]: https://github.com/Kava-Labs/kava/pull/757/commits/5b82f9a1b6374e4973ea4c8769fe1937c0fcb9e6

File: x/incentive/keeper/rewards.go (hunk 9 above)
Comment (denalimarsh) [reply x/incentive/keeper/rewards.go]: Updated in https://github.com/Kava-Labs/kava/pull/757/commits/e300ada916f042f11e21e6d0c37e49fedb388380

File: x/incentive/keeper/rewards.go (hunk 10 above)
Comment (denalimarsh) [reply x/incentive/keeper/rewards.go]: Updated in https://github.com/Kava-Labs/kava/pull/757/commits/cd59a7c5dc25240e2f45e7a9facaa9e1966f5528

File: x/incentive/keeper/rewards.go (hunk 11 above)
Comment (denalimarsh) [reply x/incentive/keeper/rewards.go:359]: Went with `if len(supplyRewardIndexes) == 0 { return }` in https://github.com/Kava-Labs/kava/pull/757/commits/cd59a7c5dc25240e2f45e7a9facaa9e1966f5528. Since `supplyRewardIndexes` contains all existing and new supply reward denoms we shouldn't set the claim/update the supply reward index if it doesn't have items. However in the case that there are existing supply reward indexes and no new supply reward index denoms, this solution will effectively set claim.SupplyRewardIndexes to itself.

File: x/incentive/keeper/rewards.go (hunk 12 above)
Comment (denalimarsh) [reply x/incentive/keeper/rewards.go]: Updated in https://github.com/Kava-Labs/kava/pull/757/commits/cd59a7c5dc25240e2f45e7a9facaa9e1966f5528

File: x/incentive/keeper/rewards.go (hunk 13 above)
Comment (denalimarsh) [reply x/incentive/keeper/rewards.go:385]: Updated in https://github.com/Kava-Labs/kava/pull/757/commits/cd59a7c5dc25240e2f45e7a9facaa9e1966f5528

**Comment by denalimarsh** [comment]
@karzak addressed the initial revisions.

**Comment by denalimarsh** [comment]
@karzak added tests for the new hooks in https://github.com/Kava-Labs/kava/pull/757/commits/a9a6c31fe5fc07ada25db38afd7b17a439e25ee9, https://github.com/Kava-Labs/kava/pull/757/commits/1e8956ddf0f02aa36ef4000de1bde1422ed52236. The tests revealed 2x little bugs which I fixed. May need to add additional test cases to cover some of the edge cases you pointed out, but should be easy to add them in if needed.

File: x/incentive/keeper/rewards.go [hunk 17]
```diff
@@ -1,66 +1,143 @@
...
+		newRewards := timeElapsed.Mul(rewardPeriod.RewardsPerSecond.Amount)
+		rewardFactor := newRewards.ToDec().Quo(totalBorrowed)
+
+		previousRewardFactor, found := k.GetHardBorrowRewardFactor(ctx, rewardPeriod.CollateralType)
+		if !found {
+			previousRewardFactor = sdk.ZeroDec()
+		}
+		newRewardFactor := previousRewardFactor.Add(rewardFactor)
+		k.SetHardBorrowRewardFactor(ctx, rewardPeriod.CollateralType, newRewardFactor)
+		k.SetPreviousHardBorrowRewardAccrualTime(ctx, rewardPeriod.CollateralType, ctx.BlockTime())
```
Comment (karzak) [inline x/incentive/keeper/rewards.go:79]: call `SetPreviousHardBorrowRewardAccrualTime` outside of if condition - we want to set the previous accrual time even if the totalBorrowedCoins is zero/not found

File: x/incentive/keeper/rewards.go [hunk 18]
```diff
@@ -1,66 +1,143 @@
...
+		return nil
+	}
+	totalBorrowedCoins, foundTotalBorrowedCoins := k.hardKeeper.GetBorrowedCoins(ctx)
+	if foundTotalBorrowedCoins {
+		totalBorrowed := totalBorrowedCoins.AmountOf(rewardPeriod.CollateralType).ToDec()
+		if totalBorrowed.IsZero() {
+			k.SetPreviousHardBorrowRewardAccrualTime(ctx, rewardPeriod.CollateralType, ctx.BlockTime())
+			return nil
+		}
+		newRewards := timeElapsed.Mul(rewardPeriod.RewardsPerSecond.Amount)
```
Comment (karzak) [inline x/incentive/keeper/rewards.go:70]: per Ruaridh's audit of usdx incentive accumulators:

```suggestion
newRewards := timeElapsed.Mul(rewardPeriod.RewardsPerSecond.Amount)
hardFactor, found := k.hardKeeper.GetInterestFactor(ctx, rewardPeriod.CollateralType)
if !found {
	k.SetPreviousHardBorrowRewardAccrualTime(ctx, rewardPeriod.CollateralType, ctx.BlockTime())
	return nil
}
rewardFactor := newRewards.ToDec().Mul(hardFactor).Quo(totalBorrowed)
```

File: x/incentive/keeper/rewards.go [hunk 19]
```diff
@@ -1,66 +1,143 @@
...
+	}
+
+	totalSuppliedCoins, foundTotalSuppliedCoins := k.hardKeeper.GetSuppliedCoins(ctx)
+	if foundTotalSuppliedCoins {
+		totalSupplied := totalSuppliedCoins.AmountOf(rewardPeriod.CollateralType).ToDec()
+		if totalSupplied.IsZero() {
+			k.SetPreviousHardSupplyRewardAccrualTime(ctx, rewardPeriod.CollateralType, ctx.BlockTime())
+			return nil
+		}
+		newRewards := timeElapsed.Mul(rewardPeriod.RewardsPerSecond.Amount)
```
Comment (karzak) [inline x/incentive/keeper/rewards.go:108]: Same as above:

```suggestion
newRewards := timeElapsed.Mul(rewardPeriod.RewardsPerSecond.Amount)
hardFactor, found := k.hardKeeper.GetInterestFactor(ctx, rewardPeriod.CollateralType)
if !found {
    k.SetPreviousHardSupplyRewardAccrualTime(ctx, rewardPeriod.CollateralType, ctx.BlockTime())
    return nil
}
rewardFactor := newRewards.ToDec().Mul(hardFactor).Quo(totalSupplied)
```

File: x/incentive/keeper/rewards.go [hunk 20]
```diff
@@ -1,66 +1,143 @@
...
+		rewardFactor := newRewards.ToDec().Quo(totalSupplied)
+
+		previousRewardFactor, found := k.GetHardSupplyRewardFactor(ctx, rewardPeriod.CollateralType)
+		if !found {
+			previousRewardFactor = sdk.ZeroDec()
+		}
+		newRewardFactor := previousRewardFactor.Add(rewardFactor)
+		k.SetHardSupplyRewardFactor(ctx, rewardPeriod.CollateralType, newRewardFactor)
+		k.SetPreviousHardSupplyRewardAccrualTime(ctx, rewardPeriod.CollateralType, ctx.BlockTime())
```
Comment (karzak) [inline x/incentive/keeper/rewards.go:117]: ditto, call `SetPreviousHardSupplyRewardAccrualTime` outside of top level if statement

**Review by karzak — CHANGES_REQUESTED** [review]
Looking good, mostly minor stuff at this point or things from Ruaridh's audit of USDX incentives that were off in the initial spec.

**Review by karzak — CHANGES_REQUESTED** [review]
Looking good, mostly minor stuff at this point or things from Ruaridh's audit of USDX incentives that were off in the initial spec.

File: app/app.go [hunk 21]
```diff
@@ -380,16 +391,6 @@ func NewApp(logger log.Logger, db dbm.DB, traceStore io.Writer, appOpts AppOptio
...
-		hardSubspace,
-		app.accountKeeper,
-		app.supplyKeeper,
-		&stakingKeeper,
-		app.pricefeedKeeper,
-		app.auctionKeeper,
-	)
 
 	// register the staking hooks
 	// NOTE: stakingKeeper above is passed by reference, so that it will contain these hooks
```
[the remark this answers is not in the document]
Comment (denalimarsh) [reply app/app.go:396]: Addressed in https://github.com/Kava-Labs/kava/pull/757/commits/c52ba54179860ad38e0219d95e1623e6a78c2ee1

File: x/hard/keeper/deposit.go [hunk 22]
```diff
@@ -20,6 +20,12 @@ func (k Keeper) Deposit(ctx sdk.Context, depositor sdk.AccAddress, coins sdk.Coi
 
 	k.SyncOutstandingInterest(ctx, depositor)
```
[the remark this answers is not in the document]
Comment (denalimarsh) [reply x/hard/keeper/deposit.go]: Addressed in https://github.com/Kava-Labs/kava/pull/757/commits/c52ba54179860ad38e0219d95e1623e6a78c2ee1

File: x/hard/keeper/deposit.go [hunk 23]
```diff
@@ -93,6 +107,10 @@ func (k Keeper) Withdraw(ctx sdk.Context, depositor sdk.AccAddress, coins sdk.Co
...
```
[the remark this answers is not in the document]
Comment (denalimarsh) [reply x/hard/keeper/deposit.go]: Addressed in https://github.com/Kava-Labs/kava/pull/757/commits/c52ba54179860ad38e0219d95e1623e6a78c2ee1

File: x/hard/keeper/repay.go [hunk 24]
```diff
@@ -25,11 +25,14 @@ func (k Keeper) Repay(ctx sdk.Context, sender sdk.AccAddress, coins sdk.Coins) e
...
 	// Check borrow exists here to avoid duplicating store read in ValidateRepay
-	borrow, found := k.GetBorrow(ctx, sender)
-	if !found {
+	borrow, hasExistingBorrow := k.GetBorrow(ctx, sender)
+	if !hasExistingBorrow {
 		return types.ErrBorrowNotFound
 	}
 
+	// Call incentive hook
+	k.BeforeBorrowModified(ctx, borrow)
```
[the remark this answers is not in the document]
Comment (denalimarsh) [reply x/hard/keeper/repay.go:31]: Addressed in https://github.com/Kava-Labs/kava/pull/757/commits/c52ba54179860ad38e0219d95e1623e6a78c2ee1

File: x/hard/keeper/borrow.go [hunk 25]
```diff
@@ -28,10 +28,12 @@ func (k Keeper) Borrow(ctx sdk.Context, borrower sdk.AccAddress, coins sdk.Coins
 		return err
 	}
 
-	// If the user has an existing borrow, sync its outstanding interest
-	_, found := k.GetBorrow(ctx, borrower)
-	if found {
-		k.SyncOutstandingInterest(ctx, borrower)
+	k.SyncOutstandingInterest(ctx, borrower)
```
[the remark this answers is not in the document]
Comment (denalimarsh) [reply x/hard/keeper/borrow.go]: Addressed in https://github.com/Kava-Labs/kava/pull/757/commits/c52ba54179860ad38e0219d95e1623e6a78c2ee1

File: x/hard/keeper/repay.go [hunk 26]
```diff
@@ -47,6 +50,9 @@ func (k Keeper) Repay(ctx sdk.Context, sender sdk.AccAddress, coins sdk.Coins) e
 	// Update total borrowed amount
 	k.DecrementBorrowedCoins(ctx, payment)
 
+	// Call incentive hook
+	k.AfterBorrowModified(ctx, borrow)
```
[the remark this answers is not in the document]
Comment (denalimarsh) [reply x/hard/keeper/repay.go]: Added the desired functionality in https://github.com/Kava-Labs/kava/pull/757/commits/c52ba54179860ad38e0219d95e1623e6a78c2ee1 so we don't forget during the rebase.

File: x/incentive/keeper/rewards.go (hunk 17 above)
Comment (denalimarsh) [reply x/incentive/keeper/rewards.go]: Updated in https://github.com/Kava-Labs/kava/pull/757/commits/801595575ed6235e156ad7e8ad08b86f348e70c6

File: x/incentive/keeper/rewards.go (hunk 18 above)
Comment (denalimarsh) [reply x/incentive/keeper/rewards.go:70]: Updated in https://github.com/Kava-Labs/kava/pull/757/commits/801595575ed6235e156ad7e8ad08b86f348e70c6

File: x/incentive/keeper/rewards.go (hunk 19 above)
Comment (denalimarsh) [reply x/incentive/keeper/rewards.go:113]: Updated in https://github.com/Kava-Labs/kava/pull/757/commits/801595575ed6235e156ad7e8ad08b86f348e70c6

File: x/incentive/keeper/rewards.go (hunk 20 above)
Comment (denalimarsh) [reply x/incentive/keeper/rewards.go]: Updated in https://github.com/Kava-Labs/kava/pull/757/commits/801595575ed6235e156ad7e8ad08b86f348e70c6

**Comment by denalimarsh** [comment]
Pushed revisions and updated the tests. Fetching the interest factor from Hard module and incorporating it in reward accumulation results in more accurate reward factors and rewards over time.

**Review by karzak — APPROVED** [review]

**Comment by denalimarsh** [comment]
Closed in favor of https://github.com/Kava-Labs/kava/pull/773
